src/Swarm/Game/Step.hs

```diff
@@ -1617,3 +1623,43 @@ safeExp :: Has (Throw Exn) sig m => Integer -> Integer -> m Integer
 safeExp a b
   | b < 0 = throwError $ CmdFailed Exp "Negative exponent"
   | otherwise = return $ a ^ b
+
+-- | Update the global list of discovered entities, and check for new recipes.
+updateDiscoveredEntities :: (Has (State GameState) sig m, Has (State Robot) sig m) => Entity -> m ()
+updateDiscoveredEntities e = do
+  allDiscovered <- use allDiscoveredEntities
+  if E.contains0plus e allDiscovered
+    then pure ()
+    else do
+      let newAllDiscovered = E.insertCount 9001 e allDiscovered
```

Comment:
`9001`?  I assume the idea is that this is supposed to be sufficiently large so that any recipe will use at most 9001 of something?  If so, I definitely don't like this on principle, though I think I understand why you did it this way.  We want to say a recipe is "available" even if you don't actually yet have *enough* of some entity, just as long as you *know about* all the inputs to the recipe, right?

How about we add a new function to the `Recipes` module, like `missingIngredientsFor`, but which just returns a `Bool` saying whether you know about all the ingredients?

Reply:
That's exactly it! I can add a new ingredients check that does not look for quantities.

Reply:
Fixed in e660c2d